### PR TITLE
Changing the way we log the version of WebViews and also handling only official WebViews from Google

### PR DESCRIPTION
**Pull request by tevincent**
Sadly, I couldn't find a way to redirect to the Huawei AppGallery in case the Huawei WebView is outdated. So we are only checking the WebView version of Google WebViews.

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt [hunk 1]
```diff
@@ -66,18 +66,34 @@ class ThreadListViewModel @Inject constructor(
...
+                    webViewMajorVersion < WEBVIEW_OFFICIAL_MIN_VERSION
+                }
+                else -> {
+                    // We'll add other package names in the future if needed here
+                    false
+                }
+            }
+        }
+
+        // TODO Remove this log in a few weeks/month if we don't have any Sentry anymore
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt:89]:
```suggestion
        // TODO: (23/07) Remove this log in a few weeks/month if we don't have any Sentry anymore
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt [hunk 2]
```diff
@@ -66,18 +66,34 @@ class ThreadListViewModel @Inject constructor(
...
+            return when (webViewPackageName) {
+                WEBVIEW_OFFICIAL_PACKAGE_NAME -> {
+                    webViewMajorVersion < WEBVIEW_OFFICIAL_MIN_VERSION
+                }
+                else -> {
+                    // We'll add other package names in the future if needed here
+                    false
+                }
+            }
+        }
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt:87]: I think we usually put the inner functions on top of the impacted scope. I don't find it really readable like this.

Comment (FabianDevel) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt:87]: I'd rewrite the entire function like this : 
```suggestion
        fun checkWebViewVersion(canShowWebViewOutdated: Boolean) {
        val webViewPackage = WebViewCompat.getCurrentWebViewPackage(appContext)
        val webViewPackageName = webViewPackage?.packageName

        val (webViewVersionName, webViewMajorVersion) = webViewPackage?.versionName?.let { versionName ->
            val majorVersion = runCatching { versionName.substringBefore('.').toInt() }.getOrDefault(defaultValue = 0)
            versionName to majorVersion
        } ?: (null to 0)

        fun hasOutdatedMajorVersion(): Boolean = when (webViewPackageName) {
                WEBVIEW_OFFICIAL_PACKAGE_NAME -> webViewMajorVersion < WEBVIEW_OFFICIAL_MIN_VERSION
                else -> false // We'll add other package names in the future if needed here
        }

        // TODO: (23/07) Remove this log in a few weeks/month if we don't have any Sentry anymore
        val hasOudatedMajorVersion = hasOutdatedMajorVersion()
        if (hasOudatedMajorVersion) SentryDebug.sendWebViewVersionName(webViewPackageName, webViewVersionName, webViewMajorVersion)
        isWebViewOutdated.value = canShowWebViewOutdated && hasOudatedMajorVersion
    }
```

Comment (FabianDevel) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt:87]: I kept the inner function here, but I'm not sure about its utility, I think we could just `val hasOutdatedVersion = when {...`

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt [hunk 3]
```diff
@@ -66,18 +66,34 @@ class ThreadListViewModel @Inject constructor(
...
 
     fun checkWebViewVersion(canShowWebViewOutdated: Boolean) {
+        val webViewPackage =  WebViewCompat.getCurrentWebViewPackage(appContext)
+        val webViewPackageName =  webViewPackage?.packageName
         val (webViewVersionName, webViewMajorVersion) =
-            WebViewCompat.getCurrentWebViewPackage(appContext)?.versionName?.let { versionName ->
+            webViewPackage?.versionName?.let { versionName ->
                 val majorVersion = runCatching { versionName.substringBefore('.').toInt() }.getOrDefault(0)
                 versionName to majorVersion
             } ?: (null to 0)
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt:75]: Magic values

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt [hunk 4]
```diff
@@ -66,18 +66,31 @@ class ThreadListViewModel @Inject constructor(
...
-                val majorVersion = runCatching { versionName.substringBefore('.').toInt() }.getOrDefault(0)
-                versionName to majorVersion
-            } ?: (null to 0)
+        val webViewPackage = WebViewCompat.getCurrentWebViewPackage(appContext)
+        val webViewPackageName = webViewPackage?.packageName
+        
+        val (webViewVersionName, webViewMajorVersion) = webViewPackage?.versionName?.let { versionName ->
+            val majorVersion = runCatching { versionName.substringBefore('.').toInt() }.getOrDefault(defaultValue = 0)
+            versionName to majorVersion
+        } ?: (null to 0)
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt:75]: Double `0` magic value.

(c) le FAB.

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt (hunk 4 above)
Comment (FabianDevel) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt]: Too late

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt [hunk 5]
```diff
@@ -66,18 +66,32 @@ class ThreadListViewModel @Inject constructor(
...
-        val (webViewVersionName, webViewMajorVersion) =
-            WebViewCompat.getCurrentWebViewPackage(appContext)?.versionName?.let { versionName ->
-                val majorVersion = runCatching { versionName.substringBefore('.').toInt() }.getOrDefault(0)
-                versionName to majorVersion
-            } ?: (null to 0)
+        val webViewPackage = WebViewCompat.getCurrentWebViewPackage(appContext)
+        val webViewPackageName = webViewPackage?.packageName
+        
+        val (webViewVersionName, webViewMajorVersion) = webViewPackage?.versionName?.let { versionName ->
+            val majorVersion = runCatching { versionName.substringBefore('.').toInt() }.getOrDefault(defaultValue = 0)
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt:73]: MAGIC NUMBER

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt (hunk 4 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt]: Nope.

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt [hunk 6]
```diff
@@ -66,18 +66,34 @@ class ThreadListViewModel @Inject constructor(
...
+        val hasOutdatedMajorVersion = when (webViewPackageName) {
+            WEBVIEW_OFFICIAL_PACKAGE_NAME -> webViewMajorVersion < WEBVIEW_OFFICIAL_MIN_VERSION
+            else -> false // We'll add other package names in the future if needed here
+        }
+
+        // TODO: (23/07) Remove this log in a few weeks/month if we don't have any Sentry anymore
+        if (hasOutdatedMajorVersion) SentryDebug.sendWebViewVersionName(
+            webViewPackageName,
+            webViewVersionName,
+            webViewMajorVersion
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt:89]:
```suggestion
            webViewMajorVersion,
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

**Review by FabianDevel — APPROVED** [review]

**Review by KevinBoulongne — APPROVED** [review]

**Pull request merged**